[css-scroll-snap] make resnap follow scroll snap target if necessary

nmoucht opened the pull request:
#### 5ed2b1dffea4b8da9c99f3e9eee1e792b66f37ef
<pre>
[css-scroll-snap] make resnap follow scroll snap target if necessary
<a href="https://bugs.webkit.org/show_bug.cgi?id=244745">https://bugs.webkit.org/show_bug.cgi?id=244745</a>
&lt;rdar://99557242&gt;

Reviewed by Martin Robinson.

CSS scroll snap spec (<a href="https://www.w3.org/TR/css-scroll-snap-1/#re-snap)">https://www.w3.org/TR/css-scroll-snap-1/#re-snap)</a>: &quot;If multiple boxes were snapped
before and their snap positions no longer coincide, then if one of them is focused or targeted, the scroll
container must re-snap to that one and otherwise which one to re-snap to is UA-defined.&quot; To acheive this,
we add an id to scroll offset which represents the associated element to that scroll offset. We also add a bool which
represents wether the associated element is focused. The id of the currently snapped element is added
to ScrollSnapAnimatorState and for each relayout, check if it is necessary to preserve the currently
snapped element.

* LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-snap/snap-after-relayout/resnap-to-focused-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-snap/snap-after-relayout/snap-to-different-targets-expected.txt:
* Source/WebCore/page/FrameView.cpp:
(WebCore::FrameView::updateSnapOffsets):
* Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp:
(WebCore::updateSnapOffsetsForScrollableArea):
(WebCore::convertOffsetInfo):
* Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.h:
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/ScrollSnapAnimatorState.cpp:
(WebCore::ScrollSnapAnimatorState::setFocusedElementForAxis):
(WebCore::ScrollSnapAnimatorState::preserveCurrentTargetForAxis):
(WebCore::ScrollSnapAnimatorState::resnapAfterLayout):
* Source/WebCore/platform/ScrollSnapAnimatorState.h:
(WebCore::ScrollSnapAnimatorState::activeSnapIDForAxis const):
(WebCore::ScrollSnapAnimatorState::setActiveSnapIndexIDForAxis):
* Source/WebCore/platform/ScrollableArea.cpp:
(WebCore::ScrollableArea::resnapAfterLayout):
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::updateSnapOffsets):
* Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingCoordinatorTransaction.cpp:
(ArgumentCoder&lt;SnapOffset&lt;float&gt;&gt;::encode):
(ArgumentCoder&lt;SnapOffset&lt;float&gt;&gt;::decode):

Canonical link: <a href="https://commits.webkit.org/254773@main">https://commits.webkit.org/254773@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f07c745b2a5fe32ad47939c8bd61f8a30d7e1e23

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/90179 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/34724 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/20791 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/99491 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/156992 "Found 1 new test failure: fast/workers/worker-crash-with-invalid-location.html (failure)") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/94188 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/33215 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/28557 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/82505 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/95968 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/95833 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/26423 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/77017 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/26308 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/81259 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/81045 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/69307 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/66/builds/34355 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/15119 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/32196 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/16067 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3352 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/63/builds/35940 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/39023 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/37843 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/35151 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.h:
```diff
@@ -38,12 +38,15 @@ namespace WebCore {
     T offset;
     ScrollSnapStop stop;
     bool hasSnapAreaLargerThanViewport;
+    unsigned snapTargetId;
```
Rather than storing this extra ID, and the associated (and tricky to maintain) renderer -> ID map, can we just compute the snap offset for the focus thing and send it along as the "preferred snap offset" or "preferred snap index"?

Any way that doesn't involve maintaining a RenderBox -> ID map would be preferable.

nmoucht replied:
The issue with storing a snap index/offset is that it will change across layout, so when `updateSnapOffsetsForScrollableArea` is called, we have no way to figure out what the corresponding index/offset is. Instead of having an extra map, would it be preferable to replace the `HashSet<const RenderBox*>` already in `RenderView` with a set of structs which hold the `RenderBox` and id?

nmoucht replied:
Actually this is not a great solution as we would have to do the work ourselves to maintain the set, due to the ids. This could of course work, but since we are maintaining the map in the same way as the set (which was on RenderView already), could you explain why this isn't preferable?

smfr reviewed (changes requested):
Tests are failing, so it seems this needs more work.

nt1m commented on Source/WebCore/platform/ScrollSnapAnimatorState.h:
```diff
@@ -98,7 +111,8 @@ class ScrollSnapAnimatorState {
     void transitionToDestinationReachedState();
-
+    void preserveCurrentTargetForAxis(ScrollEventAxis axis);
+    void setFocusedElementForAxis(ScrollEventAxis axis);
```
To make style bot happy:
```suggestion
    void preserveCurrentTargetForAxis(ScrollEventAxis);
    void setFocusedElementForAxis(ScrollEventAxis);
```

nmoucht replied:
Fixed.

mrobinson commented on Source/WebCore/platform/ScrollSnapAnimatorState.cpp:
```diff
@@ -95,18 +95,72 @@ float ScrollSnapAnimatorState::adjustedScrollDestination(ScrollEventAxis axis, F
+    auto snapOffsetsVertical = snapOffsetsForAxis(ScrollEventAxis::Vertical);
+    auto snapOffsetsHorizontal = snapOffsetsForAxis(ScrollEventAxis::Horizontal);
+    
+    // Check if we need to set the current indeces
```
```suggestion
    // Check if we need to set the current indices.
```

nmoucht replied:
Fixed.

mrobinson commented on Source/WebCore/platform/ScrollSnapAnimatorState.cpp:
```diff
@@ -95,18 +95,72 @@ float ScrollSnapAnimatorState::adjustedScrollDestination(ScrollEventAxis axis, F
+    
+    // If we have an active target, see if we need to preserve it
+    if (activeHorizontalID && activeHorizontalIndex) {
+        if (*activeHorizontalID != snapOffsetsHorizontal[*activeHorizontalIndex].snapTargetId)
```
If combined into one if statement, this would be a bit clearer, I think. For instance:

```c++
    if (activeHorizontalID && activeHorizontalIndex &&
        *activeHorizontalID != snapOffsetsHorizontal[*activeHorizontalIndex].snapTargetId)
```

nmoucht replied:
I think we want to leave the logic like this for horizontal axis. This is due to an implementation detail where we need to choose an axis to prioritize the target for (we choose horizontal). This prevents us from trying to follow the target of the vertical axis when we are already following the target of the horizontal axis.

mrobinson commented on Source/WebCore/platform/ScrollSnapAnimatorState.cpp:
```diff
@@ -95,18 +95,72 @@ float ScrollSnapAnimatorState::adjustedScrollDestination(ScrollEventAxis axis, F
+        if (*activeHorizontalID != snapOffsetsHorizontal[*activeHorizontalIndex].snapTargetId)
+            preserveCurrentTargetForAxis(ScrollEventAxis::Horizontal);
+    } else if (activeVerticalID && activeVerticalIndex) {
+        if (*activeVerticalID != snapOffsetsVertical[*activeVerticalIndex].snapTargetId)
```
Ditto.

nmoucht replied:
Fixed.

mrobinson commented on Source/WebCore/platform/ScrollSnapAnimatorState.cpp:
```diff
@@ -95,18 +95,72 @@ float ScrollSnapAnimatorState::adjustedScrollDestination(ScrollEventAxis axis, F
+    auto found = std::find_if(snapOffsets.begin(), snapOffsets.end(), [](SnapOffset<LayoutUnit> p) -> bool {
+        return p.isFocused;
+    });
+    if (found != snapOffsets.end()) {
```
I have a slight preference for an early return here.

nmoucht replied:
Fixed.

mrobinson commented on Source/WebCore/platform/ScrollSnapAnimatorState.cpp:
```diff
@@ -95,18 +95,72 @@ float ScrollSnapAnimatorState::adjustedScrollDestination(ScrollEventAxis axis, F
+    // If we have an active target, see if we need to preserve it
+    if (activeHorizontalID && activeHorizontalIndex) {
+        if (*activeHorizontalID != snapOffsetsHorizontal[*activeHorizontalIndex].snapTargetId)
+            preserveCurrentTargetForAxis(ScrollEventAxis::Horizontal);
```
Should this update `snapPointChanged` as well? I think it might need to so that the new snap index is sent to the Scrolling Tree.

nmoucht replied:
Fixed.

mrobinson commented on Source/WebCore/platform/ScrollSnapAnimatorState.cpp:
```diff
@@ -95,18 +95,72 @@ float ScrollSnapAnimatorState::adjustedScrollDestination(ScrollEventAxis axis, F
+    }
+
+    // If we are snapped to multiple targets, save them
+    if ((!activeHorizontalID && activeHorizontalIndex) && (!activeVerticalID && activeVerticalIndex) && snapOffsetsForAxis(ScrollEventAxis::Horizontal)[*activeHorizontalIndex].snapTargetId != snapOffsetsForAxis(ScrollEventAxis::Vertical)[*activeVerticalIndex].snapTargetId) {
```
This should probably join the if-else-if sequence above, just to make it clearer that this will not happen if we are preserving a previous snap point. I was a bit thrown off by this at first, which is why I mention it.

nmoucht replied:
At least to me it makes more sense to have them separate, as the first set of checks are for preserving the target for a particular axis, while the second check is for setting the targets if we don't have any. These seem logically separate to me. I updated the comment to make this difference a bit clearer.

mrobinson commented on Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.h:
```diff
@@ -38,12 +38,15 @@ namespace WebCore {
     T offset;
     ScrollSnapStop stop;
     bool hasSnapAreaLargerThanViewport;
+    unsigned snapTargetId;
```
Name nitpick: This is written as "ID" everywhere else, but "Id" here. :slightly_smiling_face:

nmoucht replied:
Fixed.

mrobinson commented on Source/WebCore/rendering/RenderView.cpp:
```diff
@@ -957,11 +957,13 @@ RenderLayer* RenderView::takeStyleChangeLayerTreeMutationRoot()
 void RenderView::registerBoxWithScrollSnapPositions(const RenderBox& box)
 {
     m_boxesWithScrollSnapPositions.add(&box);
+    m_boxesWithScrollSnapPositionsIDMap.add(&box, m_boxesWithScrollSnapPositionsCurrentID++);
```
I wonder about this approach. The specification says that resnapping after layout should be preserved according the *element*, but IIUC style change can destroy and recreate an element's `RenderView`. Is there a way to rework this so that these identifiers are preserved when the `RenderView` is recreated?

nmoucht replied:
While I'm not too familiar with the lifetime of `RenderView`, at least while testing this change I didn't see the `RenderView` being destroyed due to any style changes. However, another solution is to use the `identifier()` function on each `Element`. This also allows us to not maintain a map anymore, which Simon had some concerns about previously. This also might prevent any lifetime issues.

mrobinson replied:
Thanks. The new approach looks great to me.

mrobinson reviewed (changes requested):
Thanks for working on this tricky issue! I have a few comments, most of which are pretty minor, apart from one more serious concern about associated these new ids with the `RenderView` and not the `Element`.

nt1m commented on Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp:
```diff
@@ -277,7 +277,7 @@ static std::pair<bool, bool> axesFlippedForWritingModeAndDirection(WritingMode w
 }
 
-void updateSnapOffsetsForScrollableArea(ScrollableArea& scrollableArea, const RenderBox& scrollingElementBox, const RenderStyle& scrollingElementStyle, LayoutRect viewportRectInBorderBoxCoordinates, WritingMode writingMode, TextDirection textDirection)
+void updateSnapOffsetsForScrollableArea(ScrollableArea& scrollableArea, const RenderBox& scrollingElementBox, const RenderStyle& scrollingElementStyle, LayoutRect viewportRectInBorderBoxCoordinates, WritingMode writingMode, TextDirection textDirection, Element* focuesedElement)
```
typo: 
```suggestion
void updateSnapOffsetsForScrollableArea(ScrollableArea& scrollableArea, const RenderBox& scrollingElementBox, const RenderStyle& scrollingElementStyle, LayoutRect viewportRectInBorderBoxCoordinates, WritingMode writingMode, TextDirection textDirection, Element* focusedElement)
```

nmoucht replied:
Fixed.

mrobinson reviewed (approved):
This looks good to me with the changes that @nt1m suggests. Thanks!

nt1m commented on Source/WebCore/platform/ScrollSnapAnimatorState.h:
```diff
@@ -112,6 +126,9 @@ class ScrollSnapAnimatorState {
     std::optional<unsigned> m_activeSnapIndexY;
+    std::optional<unsigned> m_activeSnapIDX;
+    std::optional<unsigned> m_activeSnapIDY;
+    
```
nit: superfluous new line

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/platform/ScrollSnapAnimatorState.cpp:
```diff
@@ -95,18 +95,72 @@ float ScrollSnapAnimatorState::adjustedScrollDestination(ScrollEventAxis axis, F
+        setActiveSnapIndexIDForAxis(ScrollEventAxis::Horizontal, snapOffsetsForAxis(ScrollEventAxis::Horizontal)[*activeHorizontalIndex].snapTargetID);
+        setActiveSnapIndexIDForAxis(ScrollEventAxis::Vertical, snapOffsetsForAxis(ScrollEventAxis::Vertical)[*activeVerticalIndex].snapTargetID);
+    }
+    LOG_WITH_STREAM(ScrollSnap, stream << "ScrollSnapAnimatorState::resnapAfterLayout() current target: horizontal" << activeHorizontalID << " vertical: "<< activeVerticalID);
```
nit: the logging could be more readable with this:
```suggestion
    LOG_WITH_STREAM(ScrollSnap, stream << "ScrollSnapAnimatorState::resnapAfterLayout() current target: horizontal: " << activeHorizontalID << " vertical: "<< activeVerticalID);
```

nmoucht replied:
Fixed.

nmoucht committed:
[css-scroll-snap] make resnap follow scroll snap target if necessary
https://bugs.webkit.org/show_bug.cgi?id=244745
<rdar://99557242>

Reviewed by Martin Robinson.

CSS scroll snap spec (https://www.w3.org/TR/css-scroll-snap-1/#re-snap): "If multiple boxes were snapped
before and their snap positions no longer coincide, then if one of them is focused or targeted, the scroll
container must re-snap to that one and otherwise which one to re-snap to is UA-defined." To acheive this,
we add an id to scroll offset which represents the associated element to that scroll offset. We also add a bool which
represents wether the associated element is focused. The id of the currently snapped element is added
to ScrollSnapAnimatorState and for each relayout, check if it is necessary to preserve the currently
snapped element.

* LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-snap/snap-after-relayout/resnap-to-focused-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-snap/snap-after-relayout/snap-to-different-targets-expected.txt:
* Source/WebCore/page/FrameView.cpp:
(WebCore::FrameView::updateSnapOffsets):
* Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp:
(WebCore::updateSnapOffsetsForScrollableArea):
(WebCore::convertOffsetInfo):
* Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.h:
(WebCore::operator<<):
* Source/WebCore/platform/ScrollSnapAnimatorState.cpp:
(WebCore::ScrollSnapAnimatorState::setFocusedElementForAxis):
(WebCore::ScrollSnapAnimatorState::preserveCurrentTargetForAxis):
(WebCore::ScrollSnapAnimatorState::resnapAfterLayout):
* Source/WebCore/platform/ScrollSnapAnimatorState.h:
(WebCore::ScrollSnapAnimatorState::activeSnapIDForAxis const):
(WebCore::ScrollSnapAnimatorState::setActiveSnapIndexIDForAxis):
* Source/WebCore/platform/ScrollableArea.cpp:
(WebCore::ScrollableArea::resnapAfterLayout):
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::updateSnapOffsets):
* Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingCoordinatorTransaction.cpp:
(ArgumentCoder<SnapOffset<float>>::encode):
(ArgumentCoder<SnapOffset<float>>::decode):

Canonical link: https://commits.webkit.org/254773@main
